ENH: Add Series.str.casefold

charlesdong1991 opened the pull request:
- [x] closes #25405  
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

charlesdong1991 committed:
collect updated master

charlesdong1991 committed:
add whatsnew

charlesdong1991 committed:
small change

gfyoung commented on pandas/tests/test_strings.py:
```diff
@@ -3424,3 +3424,11 @@ def test_method_on_bytes(self):
+        s = Series(['ß', NA, 'case', 'ßd'])
+        exp = s.str.casefold()
+
+        assert isinstance(exp, Series)
```
This line is unnecessary.  `tm.assert_series_equal` will take care of that.

gfyoung commented on pandas/tests/test_strings.py:
```diff
@@ -3424,3 +3424,11 @@ def test_method_on_bytes(self):
+    def test_casefold(self):
+        values = Series(['ss', NA, 'case', 'ssd'])
+        s = Series(['ß', NA, 'case', 'ßd'])
+        exp = s.str.casefold()
```
`exp` is misleading as a variable name, as this isn't the "expected output".

Something like `result` would be more suitable, I think.

charlesdong1991 replied:
thanks, @gfyoung , changed!

charlesdong1991 committed:
remove unnecessary test

charlesdong1991 committed:
rename

gfyoung reviewed (approved):
cc @jreback

jreback commented:
needs a mention in text.rst (also there is a list of functions ) and in the references 
and a whatsnew note

jreback reviewed (changes requested):
see comments

charlesdong1991 committed:
add series.str.casefold in reference

charlesdong1991 committed:
add reference in text rst

charlesdong1991 commented:
Thanks, @jreback I added it to `text.rst` as well as to `reference/series.rst`, I am not very sure which whatsnew note you refer to, I already added one in `0.25.0.rst`, please let me know if you want it to move somewhere else.

charlesdong1991 committed:
add skipif to avoid failure

charlesdong1991 committed:
add issue number

pep8speaks commented:
Hello @charlesdong1991! Thanks for updating the PR.



 - In the file [`pandas/tests/test_strings.py`](https://github.com/pandas-dev/pandas/blob/983332ea448d859113be622cb97f8d486e50611e/pandas/tests/test_strings.py), following are the PEP8 issues :

> [Line 3430:9](https://github.com/pandas-dev/pandas/blob/983332ea448d859113be622cb97f8d486e50611e/pandas/tests/test_strings.py#L3430): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

charlesdong1991 committed:
fix pep8

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -22,8 +22,8 @@ Other Enhancements
 - Indexing of ``DataFrame`` and ``Series`` now accepts zerodim ``np.ndarray`` (:issue:`24919`)
 - :meth:`Timestamp.replace` now supports the ``fold`` argument to disambiguate DST transition times (:issue:`25017`)
 - :meth:`DataFrame.at_time` and :meth:`Series.at_time` now support :meth:`datetime.time` objects with timezones (:issue:`24043`)
+- Add ``casefold`` to ``Series.str`` (:issue:`25405`)
```
`Series.str` has gained the ``:meth:`Series.str.casefold` `` method to ......(fill this in)

charlesdong1991 replied:
changed

jreback commented on pandas/core/strings.py:
```diff
@@ -2995,6 +2996,7 @@ def rindex(self, sub, start=0, end=None):
     _shared_docs['capitalize'] = dict(type='be capitalized',
                                       method='capitalize')
     _shared_docs['swapcase'] = dict(type='be swapcased', method='swapcase')
+    _shared_docs['casefold'] = dict(type='be casefolded', method='casefold')
```
can you add the versionadded 0.25.0 here (may need to add it to the dict to be formatted)

charlesdong1991 replied:
thanks, @jreback added!

jreback replied:
can u verify this renders ok in the terminal

charlesdong1991 replied:
ahh, yeah, I printed the docstring, looks fine for me. @jreback 
```python
Convert strings in the Series/Index to be casefolded.

Equivalent to :meth:`str.casefold`.

Returns
-------
Series/Index of objects

See Also

Series.str.lower : Converts all characters to lowercase.
Series.str.upper : Converts all characters to uppercase.
Series.str.title : Converts first character of each word to uppercase and
remaining to lowercase.
Series.str.capitalize : Converts first character to uppercase and
remaining to lowercase.
Series.str.swapcase : Converts uppercase to lowercase and lowercase to
uppercase
Series.str.casefold: Removes all case distinctions in the string.
   .. versionadded:: 0.25.0
```

WillAyd replied:
Does this render well on the web? Not aware of any other instance where the version added is in the See Also section. May be worth messing around with substitution to put it in the Summary

charlesdong1991 replied:
this rendering looks fine for me in the terminal. Since the summary of rendering only allows two parameters, and among these methods, only this `Series.str.casefold` is added in the new version. So i put this in the `See also` under `Series.str.casefold`. Otherwise, i assume the rendering might need to be changed.

WillAyd replied:
You should be able to generate one HTML file rather easily:

https://python-sprints.github.io/pandas/guide/pandas_pr.html#visual-validation-of-the-docstring

If you can double check there would be preferable as majority of users will interface to docs via HTML.

charlesdong1991 replied:
well, do you have any idea why it's complaining `AttributeError: type object 'StringMethods' has no attribute 'casefold'`? is this quoting master? @WillAyd

charlesdong1991 replied:
btw, i tried something like below to put version added in the summary, but when i kind of manually inserting a blank line for other methods since they don't have version added issue. I would like to hear your advice on how to improve this? @WillAyd 
```
    _shared_docs['casemethods'] = ("""
    Convert strings in the Series/Index to %(type)s.

    Equivalent to :meth:`str.%(method)s`.
    %(version)s
```

WillAyd replied:
What are you running to get the AttributeError? As far as the second comment, probably easiest to add a newline(s) into the version argument

charlesdong1991 committed:
changes based on review

charlesdong1991 committed:
fix conflict

jreback commented on doc/source/user_guide/text.rst:
```diff
@@ -618,3 +618,4 @@ Method Summary
     :meth:`~Series.str.istitle`;Equivalent to ``str.istitle``
     :meth:`~Series.str.isnumeric`;Equivalent to ``str.isnumeric``
     :meth:`~Series.str.isdecimal`;Equivalent to ``str.isdecimal``
+    :meth:`~Series.str.casefold`;Equivalent to ``str.casefold``
```
not sure if these are ordered here - if they r alphabetic then this is not in the right place; otherwise maybe put next to lower

charlesdong1991 replied:
ok, the methods in `series.rst` is alphabetically ordered, here is not. But I will move it up a bit!

jreback reviewed (changes requested):
lgtm with some minor comments ping on green

charlesdong1991 commented:
sorry, @jreback somehow, I cannot reply to your second comment, could you explain a little bit more how to verify it? I did pytest in the terminal, and if rendering is wrong, i get error, but here i see it seems fine?

charlesdong1991 committed:
move position up

charlesdong1991 committed:
minor

WillAyd commented on pandas/tests/test_strings.py:
```diff
@@ -3424,3 +3424,12 @@ def test_method_on_bytes(self):
+    @pytest.mark.skipif(compat.PY2, reason='not in python2')
+    def test_casefold(self):
+        # GH25405
+        casefolded = Series(['ss', NA, 'case', 'ssd'])
```
Just call this `expected`

charlesdong1991 replied:
thanks, @WillAyd i just changed!

WillAyd reviewed (changes requested):
Very minor nit but otherwise lgtm

charlesdong1991 committed:
minor change on naming convention

charlesdong1991 committed:
new args for version

charlesdong1991 commented:
i change it a bit, now it looks like, do you prefer this? @WillAyd 
```
    Convert strings in the Series/Index to casefolded.
       .. versionadded:: 0.25.0
    Equivalent to :meth:`str.casefold`.

    Returns
    -------
    Series/Index of objects

    See Also
    --------
    Series.str.lower : Converts all characters to lowercase.
    Series.str.upper : Converts all characters to uppercase.
    Series.str.title : Converts first character of each word to uppercase and
        remaining to lowercase.
    Series.str.capitalize : Converts first character to uppercase and
        remaining to lowercase.
    Series.str.swapcase : Converts uppercase to lowercase and lowercase to
        uppercase.
    Series.str.casefold: Removes all case distinctions in the string.
```

jreback commented:
lgtm. over to you @WillAyd

WillAyd commented:
@charlesdong1991 does that render correctly in HTML? Might need a blank line before versionadded

charlesdong1991 commented:
somehow, i still get this error when running `python make.py html --single pandas.Series.str.casefold`, is it because of `str.casefold` not in python2? @WillAyd  
```
Traceback (most recent call last):
  File "make.py", line 79, in _process_single_doc
    obj = getattr(obj, name)
AttributeError: type object 'StringMethods' has no attribute 'casefold'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "make.py", line 339, in <module>
    sys.exit(main())
  File "make.py", line 334, in main
    args.verbosity, args.warnings_are_errors)
  File "make.py", line 46, in __init__
    single_doc = self._process_single_doc(single_doc)
  File "make.py", line 81, in _process_single_doc
    raise ImportError('Could not import {}'.format(single_doc))
ImportError: Could not import pandas.Series.str.casefold
```

WillAyd commented:
Not sure the distinction you are trying to make with Python2 - are you actively developing with that?

Otherwise do the other methods produce something for you? I would think should work the same as the rest. As far as the blank lines go can just add a "\n" preceding the argument

charlesdong1991 commented:
i distinguish python 2 is because `str.casefold` is only valid in python3+ if i am not wrong.

It's quite wired that all other methods work fine, but this new added one...  @WillAyd

charlesdong1991 committed:
add \n

WillAyd commented:
Looks like an issue in the make file which is causing it to pick up another pandas installation on your computer which obviously wouldn't have that attribute yet. I patched locally just to make sure so here's the doc:

<img width="791" alt="image" src="https://user-images.githubusercontent.com/609873/53599302-b79b9400-3b5b-11e9-8ddd-69ed032c9350.png">

lgtm

WillAyd commented:
Thanks @charlesdong1991 nice work!
